Applications/Spire/Include/Spire/Ui/AssociativeValueModel.hpp

```diff
@@ -0,0 +1,232 @@
+      [=] (auto is_selected) { on_current(value, is_selected); }));
+    if(get_state() == QValidator::Invalid) {
+      set_current(value);
+    }
+  }
+
+  template<typename T>
+  std::shared_ptr<BooleanModel> AssociativeValueModel<T>::make_association(
+      const Type& value) {
+    auto model = std::make_shared<LocalBooleanModel>(false);
+    associate(model, value);
+    return model;
+  }
+
+  template<typename T>
+  std::shared_ptr<BooleanModel> AssociativeValueModel<T>::find(
+      const Type& value) const {
+    auto iterator = m_models.find(value);
+    if(iterator == m_models.end()) {
+      return nullptr;
+    }
+    return iterator->second;
+  }
+
+  template<typename T>
+  QValidator::State AssociativeValueModel<T>::get_state() const {
+    if(m_models.find(m_current) != m_models.end()) {
+      return QValidator::Acceptable;
+    }
+    return QValidator::Invalid;
+  }
+
+  template<typename T>
+  const typename AssociativeValueModel<T>::Type&
+      AssociativeValueModel<T>::get_current() const {
+    return m_current;
+  }
+
+  template<typename T>
+  QValidator::State AssociativeValueModel<T>::set_current(const Type& value) {
+    if(m_models.find(value) == m_models.end()) {
+      return QValidator::Invalid;
+    }
+    auto previous = m_current;
+    m_current = value;
+    set_associated_model_value(previous, false);
+    set_associated_model_value(m_current, true);
+    m_current_signal(m_current);
+    return QValidator::Acceptable;
+  }
+
+  template<typename T>
+  boost::signals2::connection AssociativeValueModel<T>::connect_current_signal(
+      const typename CurrentSignal::slot_type& slot) const {
+    return m_current_signal.connect(slot);
+  }
+
+  template<typename T>
+  void AssociativeValueModel<T>::set_associated_model_value(const Type& value,
+      bool model_value) {
+    if(auto iterator = m_models.find(value); iterator != m_models.end()) {
+      iterator->second->set_current(model_value);
+    }
+  }
+
+  template<typename T>
+  void AssociativeValueModel<T>::on_current(const Type& value,
+      bool is_selected) {
+    if(value == m_current && is_selected ||
+        value != m_current && !is_selected) {
+      return;
+    }
+    if(m_is_blocked) {
+      if(is_selected) {
+        set_associated_model_value(value, false);
+      }
+      return;
+    }
+    if(m_current == value && !is_selected) {
+      if(m_default_value) {
+        m_is_blocked = true;
+        set_current(*m_default_value);
+        m_is_blocked = false;
+      } else {
+        set_associated_model_value(m_current, true);
+      }
+      return;
+    }
+    m_is_blocked = true;
+    set_current(value);
+    m_is_blocked = false;
+  }
+
+  template<typename T>
+  std::shared_ptr<AssociativeValueModel<boost::optional<T>>>
+      make_nullable_associative_model() {
+    auto model =
+      std::make_shared<AssociativeValueModel<boost::optional<T>>>(
+        boost::optional<T>(boost::none));
```

Comment:
```boost::none``` should be sufficient, no need to wrap it.

Reply:
Why was this resolved?

Reply:
I might have made some changes and reverted them after I resolved this. It was a mistake either way.